runtime/lua/vim/lsp/util.lua

```diff
@@ -226,6 +226,166 @@ end
+--@param start_char integer First different character idx of range
+--@param end_char integer Last different character idx of range
+--@param start_line integer First different line idx of range
+--@param end_line integer Last different line idx of range
+--@returns string text extracted from defined region
+local function extract_text(lines, start_line, start_char, end_line, end_char)
+	if start_line == #lines + end_line + 1 then
+		if end_line == 0 then return '' end
+		local line = lines[start_line]
+		local length = #line + end_char - start_char
+		return string.sub(line, start_char, start_char + length + 1)
+	end
+	local result = string.sub(lines[start_line], start_char) .. '\n'
+	for line_idx = start_line + 1, #lines + end_line do
+		result = result .. lines[line_idx] .. '\n'
+	end
+	if end_line ~= 0 then
+		local line = lines[#lines + end_line + 1]
+		local length = #line + end_char + 1
+		result = result .. string.sub(line, 1, length)
+	end
+	return result
+end
+
+--@private
+--- Compute the length of the substituted range
+--@param lines table list of lines
+--@param start_char integer First different character idx of range
+--@param end_char integer Last different character idx of range
+--@param start_line integer First different line idx of range
+--@param end_line integer Last different line idx of range
+--@returns (int, int) end_line_idx and end_col_idx of range
+local function compute_length(lines, start_line, start_char, end_line, end_char)
+	local adj_end_line = #lines + end_line + 1
+	local adj_end_char
+	if adj_end_line > #lines then
+		adj_end_char =  end_char - 1
+	else
+		adj_end_char = #lines[adj_end_line] + end_char
+	end
+	if start_line == adj_end_line then
+		print(adj_end_char, start_char)
+		return adj_end_char - start_char + 1
+	end
+	local result = #lines[start_line] - start_char + 1
+	for line = start_line + 1, adj_end_line -1 do
+		result = result + #lines[line] + 1
+	end
+	result = result + adj_end_char + 1
+	return result
+end
+
+--- Returns the range table for the difference between old and new lines
+--@param old_lines table list of lines
+--@param new_lines table list of lines
+--@returns table start_line_idx and start_col_idx of range
+function M.compute_diff(old_lines, new_lines, start_line_idx, end_line_idx)
+	local start_line, start_char = first_difference(old_lines, new_lines, start_line_idx)
+  local end_line, end_char = last_difference(vim.list_slice(old_lines, start_line, #old_lines, end_line_idx),
+      vim.list_slice(new_lines, start_line, #new_lines), start_char, end_line_idx)
+  local text = extract_text(new_lines, start_line, start_char, end_line, end_char)
+  local length = compute_length(old_lines, start_line, start_char, end_line, end_char)
+
+  local adj_end_line = #old_lines + end_line
+  local adj_end_char
+	if end_line == 0 then
+		adj_end_char = 0
+	else
+		adj_end_char = #old_lines[#old_lines + end_line + 1] + end_char + 1
+	end
+
+  local result = {
+		range = {
```

Comment:
I think there's a mix of tabs and spaces in here. At least on github the indentations here look rather crazy.

Reply:
Ah good catch, this is when I get for disabling vim-sleuth. Should be fixed.